File: lib/mercuryapi-1.23.0.20/java/samples_nb/src/samples/Firmware.java
/*
 * Sample program that flash the firmware on module
 * 
 */
package samples;

import com.thingmagic.FirmwareLoadOptions;
import com.thingmagic.LLRPFirmwareLoadOptions;
import com.thingmagic.LLRPReader;
import com.thingmagic.ReadListener;
import com.thingmagic.Reader;
import com.thingmagic.ReaderException;
import com.thingmagic.RqlFirmwareLoadOptions;
import com.thingmagic.RqlReader;
import com.thingmagic.SerialReader;
import com.thingmagic.TMConstants;
import com.thingmagic.TagReadData;
import java.io.FileInputStream;
import java.io.InputStream;


public class Firmware {   
      
  static void usage()
  {
    System.out.printf("Usage: demo reader-uri <command> [args]\n" +
                      "  (URI: 'tmr:///COM1' or 'tmr://astra-2100d3/' " +
                      "or 'tmr:///dev/ttyS0')\n\n" +
                      "Available commands:\n "
                     +"Fixed reader: tmr://astra-2100d3 --fw filename --erase --revert"
                     +"\n Serial reader: tmr:///COM1 --fw filename"
                     +"\n filename  name of firmware file/path to load "
                     +"\n --erase   erase contents before installing"
                     +"\n --revert  revert to default settings");
  }

  public static void setTrace(Reader r, String args[])
  {    
    if (args[0].toLowerCase().equals("on"))
    {
      r.addTransportListener(r.simpleTransportListener);
    }    
  }
   

  static class TagReadListener implements ReadListener
  {
        public void tagRead(Reader r, TagReadData t) {
            System.out.println("Tag Read " + t);
        }
  }
  
  public static void main(String argv[])
  {
        try
        {
            // Program setup
            Reader r = null;
            
            boolean trace = false;
            FileInputStream fileStream;
            boolean erase = false;
            boolean revert = false;
            boolean argError = false;
            String fwFilename = null;
            String readerUri = null;
            String arg = null;
            
            if (argv.length < 1)
            {
                usage();
            }
            
            for (int nextarg = 0; nextarg < argv.length; nextarg++) 
            {
                arg = argv[nextarg];
                if (arg.equalsIgnoreCase("-v"))
                {
                    trace = true;                    
                }
                else if (arg.equalsIgnoreCase("--erase"))
                {
                    erase = true;                    
                }
                else if (arg.equalsIgnoreCase("--revert")) 
                {
                    revert = true;
                }
                else if (arg.equalsIgnoreCase("--fw")) 
                {
                    fwFilename = argv[++nextarg];                    
                }
                else if (readerUri == null)
                {
                    readerUri = arg;                    
                }                
                else 
                {
                    System.out.println("Argument " + arg + " is not recognized");
                    argError = true;                    
                }                
            }
            if (readerUri == null) 
            {
                System.out.println("Reader URI not specified");                
                argError = true;
            }
            if (fwFilename == null) 
            {
                System.out.println("Firmware filename not specified");                
                argError = true;
            }
            if (argError) 
            {
                usage();
                System.exit(1);
            }

            // Create Reader object, connecting to physical device       
            r = Reader.create(readerUri);
            if (trace) 
            {
                setTrace(r, new String[]{"on"});
            }
            try
            {                
                r.connect();                
                if (Reader.Region.UNSPEC == (Reader.Region) r.paramGet("/reader/region/id"))
                {
                    Reader.Region[] supportedRegions = (Reader.Region[]) r.paramGet(TMConstants.TMR_PARAM_REGION_SUPPORTEDREGIONS);
                    if (supportedRegions.length < 1)
                    {
                        throw new Exception("Reader doesn't support any regions");
                    } else 
                    {
                        r.paramSet("/reader/region/id", supportedRegions[0]);
                    }
                }
            } 
            catch (Exception ex) 
            {
                if (ex.getMessage().equalsIgnoreCase("Application image failed CRC check")) 
                {
                    System.out.println("Current image failed CRC check. Replacing with new firmware");                    
                }
                else
                {
                    throw new Exception(ex);                    
                }
            }
            
            if (r instanceof RqlReader || r instanceof LLRPReader) 
            {
                fileStream = new FileInputStream(fwFilename);
                FirmwareLoadOptions flo = new LLRPFirmwareLoadOptions(erase, revert);
                System.out.println("Loading firmware...");
                r.firmwareLoad(fileStream, flo);
            } 
            else if (r instanceof SerialReader) 
            {
                fileStream = new FileInputStream(fwFilename);
                System.out.println("Loading firmware...");
                r.firmwareLoad(fileStream);
                fileStream.close();
            }
            String version = (String) r.paramGet("/reader/version/software");
            System.out.println("Firmware load successful with version " + version);
            // Shut down reader
            r.destroy();
        } 
        catch (ReaderException re) 
        {
            if (re.getMessage().equals("Invalid firmware load arguments")) 
            {
                usage();
                System.exit(1);
            } 
            else 
            {
                System.out.println("Reader Exception : " + re.getMessage());
            }
        } 
        catch (Exception re)
        {
            System.out.println("Exception : " + re.getMessage());
        }
    }    
}

File: lib/mercuryapi-1.23.0.20/cs/Samples/Codelets/Firmware/Firmware.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;
using System.Threading;

// Reference the API
using ThingMagic;

namespace Firmware
{
    class Firmware
    {
        static void usage()
        {
            Console.WriteLine(String.Join("\r\n", new string[] {
                    "Please provide valid arguments, such as:",
                    "Fixed reader: tmr://astra-2100d3 --fw filename --erase --revert",
                    "Serial reader:tmr:///COM1 --fw filename",
                   "filename -- Name of firmware file to load \n",
                }));
            Environment.Exit(1);
        }
        static void Main(string[] args)
        {
            String readerUri = null;
            String fwFilename = null;
            Boolean erase = false;
            Boolean revert = false;
            FileStream fileStream;
            Boolean argError = false;

            for (int nextarg = 0; nextarg < args.Length; nextarg++)
            {
                string arg = args[nextarg];
                if (arg.Equals("--erase"))
                {
                    erase = true;
                }
                else if (arg.Equals("--revert"))
                {
                    revert = true;
                }
                else if (arg.Equals("--fw"))
                {
                    fwFilename = args[++nextarg];
                }
                else if (null == readerUri)
                {
                    readerUri = arg;
                }
                else
                {
                    Console.WriteLine("Argument {0}:\"{1}\" is not recognized", nextarg, arg);
                    argError = true;
                }
            }
            if (null == readerUri)
            {
                Console.WriteLine("Reader URI not specified");
                argError = true;
            }

            if (null == fwFilename)
            {
                Console.WriteLine("Firmware filename not specified");
                argError = true;
            }

            if (argError)
            {
                usage();
            }

            try
            {
                // Create Reader object, connecting to physical device.
                // Wrap reader in a "using" block to get automatic
                // reader shutdown (using IDisposable interface).
                using (Reader r = Reader.Create(readerUri))
                {
                    //Uncomment this line to add default transport listener.
                    //r.Transport += r.SimpleTransportListener;
                    try
                    {
                        r.Connect();
                        Reader.Region Region = (Reader.Region)r.ParamGet("/reader/region/id");
                        Reader.Region[] getRegion = (Reader.Region[])r.ParamGet("/reader/region/supportedRegions");

                        if (Reader.Region.UNSPEC == (Reader.Region)r.ParamGet("/reader/region/id"))
                        {
                            r.ParamSet("/reader/region/id", Reader.Region.PRC);
                        }
                    }
                    catch (FAULT_BL_INVALID_IMAGE_CRC_Exception)
                    {
                        Console.WriteLine("Error: Current image failed CRC check. Replacing with new firmware...");
                    }
                    catch (FAULT_BL_INVALID_APP_END_ADDR_Exception)
                    {
                        Console.WriteLine("The last word of the firmware image stored in the reader's "
                                      + "flash ROM does not have the correct address value.");
                    }
                    if (r is RqlReader || r is LlrpReader)
                    {
                        fileStream = new FileStream(fwFilename, FileMode.Open, FileAccess.Read);
                        FirmwareLoadOptions flo = new LlrpFirmwareLoadOptions(erase, revert);
                        Console.WriteLine("Loading Firmware....");
                        r.FirmwareLoad(fileStream, flo);
                    }
                    else if (r is SerialReader)
                    {
                        fileStream = new FileStream(fwFilename, FileMode.Open, FileAccess.Read);
                        Console.WriteLine("Loading Firmware....");
                        r.FirmwareLoad(fileStream);
                    }
                    string version = (string)r.ParamGet("/reader/version/software");
                    Console.WriteLine("Firmware load successful with version {0}", version);
                }
            }

            catch (ReaderException re)
            {
                Console.WriteLine("Error: " + re.Message);
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error: " + ex.Message);
            }
        }
    }
}

